docs/proposal/2022-09-28-event-manager-extensibility.md

```diff
@@ -0,0 +1,302 @@
+helps us to communicate with this sub process. In Kubernetes environment, it will be in same pod. No need to worry about size, since only the required plugins will be donwloaded
+to local cache (user's home folder/.Botkube-plugins).
+
+### Example
+You can see a working example [here](https://github.com/huseyinbabal/Botkube-plugins-playground)
+
+### Consequences
+If we agree on this design, we can extend this epic with following description to come up with new tasks.
+
+- [ ] Maintain Botkube plugins under Kubeshop organization
+  - There is a working version of [botkube-plugins](https://github.com/huseyinbabal/botkube-plugins), move it to Kubeshop Github organization
+  - Check module names in the [botkube-plugins](https://github.com/huseyinbabal/botkube-plugins) project and set repo owner parts as `kubeshop` after migrating to Kubeshop organization.
+
+- [ ] Add plugin contribution documentation
+  - Prepare a documentation for adding new plugin to Botkube Plugins.
+  - It should contain how to verify plugin functionality in PR checks.
+  - Boilerplate template for plugin development. Please refer [here](https://github.com/huseyinbabal/botkube-plugins/tree/main/plugins/kubectl) for an example.
+  - Once needed, provide a mocking mechanism to write tests for specific plugin.
+  
+- [ ] Add plugin manager to Botkube
+  - Plugin manager can accept plugin configuration with name and version.
+  - It should download plugin executable by using name and version to specified local folder before running Botkube application
+  - There should be a caching mechanism to not download same version again.
+  - You can see working example [here](https://github.com/huseyinbabal/botkube-plugins-playground/blob/main/plugin/manager.go)
+
+- [ ] Extract Kubernetes source feature as a plugin 
+  - Currently, Kubernetes source events are tightly coupled with Botkube, we can add sourcing mechanism as `kubernetes` plugin. You can see independent example [here](https://github.com/huseyinbabal/botkube-plugins/tree/main/plugins/kubernetes)
+  - It should be background compatible.
+  - Plugin specific parameters should be passed during plugin initialization. Hashicorp's Go plugin system, runs plugin executable as sub process, so it also accepts exec arguments where we can resolve configuration and pass them as separate arguments
+  - Plugin parameters will be documented in [botkube-plugins](https://github.com/huseyinbabal/botkube-plugins) repo, and they can be provided from Botkube as an environment variable like `KUBERNETES_...` so that they can be passed to executable [in this line](https://github.com/huseyinbabal/botkube-plugins-playground/blob/c85cb7b84296a2f41c2dcdd8cac77c0e9dd9c69a/plugin/manager.go#L159)
+  - Data structure for events
+    - Since each source plugin has different data structure, we cannot have a general data structure except we specified #Cloud Events. `data` field has free data structure, but we can convert them to map then extract 
```

Comment:
but the questions is whether we really need to accept a free form? We can enforce a layout of the `data` but not necessary the naming. Sth like:

```yaml
data:
  fields:
     {field_name}: {field_value}
```

and tell them that how they will be processed by the Botkube, so they know how to shape them to achieve a given result. This can also be done as a part of the first implementation to see how this comes together.

Reply:
As we discussed in the meeting, we can use flat data structure for now, we will be testing interactive message data structure and if it works smoothly, we can use that format.